src/main/java/bio/terra/service/dataset/flight/update/DatasetSchemaUpdateAddColumnsBigQueryStep.java

```diff
@@ -0,0 +1,93 @@
+import bio.terra.service.tabulardata.google.bigquery.BigQueryDatasetPdao;
+import bio.terra.service.tabulardata.google.bigquery.BigQueryPdao;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import com.google.cloud.bigquery.BigQuery;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+public class DatasetSchemaUpdateAddColumnsBigQueryStep implements Step {
+  private final BigQueryDatasetPdao bigQueryDatasetPdao;
+  private final DatasetDao datasetDao;
+  private final UUID datasetId;
+  private final DatasetSchemaUpdateModel updateModel;
+
+  public DatasetSchemaUpdateAddColumnsBigQueryStep(
+      BigQueryDatasetPdao bigQueryDatasetPdao,
+      DatasetDao datasetDao,
+      UUID datasetId,
+      DatasetSchemaUpdateModel updateModel) {
+    this.bigQueryDatasetPdao = bigQueryDatasetPdao;
+    this.datasetDao = datasetDao;
+    this.datasetId = datasetId;
+    this.updateModel = updateModel;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    Dataset dataset = datasetDao.retrieve(datasetId);
+
+    BigQueryProject bigQueryProject = BigQueryProject.from(dataset);
+    BigQuery bigQuery = bigQueryProject.getBigQuery();
+    String datasetName = BigQueryPdao.prefixName(dataset.getName());
+
+    for (var columnChanges : updateModel.getChanges().getAddColumns()) {
+      DatasetTable table = dataset.getTableByName(columnChanges.getTableName()).orElseThrow();
+      for (ColumnModel columnModel : columnChanges.getColumns()) {
+        Column column = DatasetJsonConversion.columnModelToDatasetColumn(columnModel, List.of());
+        bigQueryDatasetPdao.createColumn(datasetName, bigQuery, table, column);
+      }
+      bigQuery.update(
+          BigQueryDatasetPdao.buildLiveView(bigQueryProject.getProjectId(), datasetName, table));
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    Dataset dataset = datasetDao.retrieve(datasetId);
+
+    BigQueryProject bigQueryProject = BigQueryProject.from(dataset);
+    BigQuery bigQuery = bigQueryProject.getBigQuery();
+    String datasetName = BigQueryPdao.prefixName(dataset.getName());
+
+    for (var columnChanges : updateModel.getChanges().getAddColumns()) {
+      DatasetTable table = dataset.getTableByName(columnChanges.getTableName()).orElseThrow();
+      Set<String> columnNames =
+          columnChanges.getColumns().stream().map(ColumnModel::getName).collect(Collectors.toSet());
+      table.columns(
+          table.getColumns().stream()
+              .filter(c -> !columnNames.contains(c.getName()))
+              .collect(Collectors.toList()));
+
+      // We can't actually drop columns from a BigQuery table. All we can do is update the live view
```

Comment:
Let's say we have to undo this step, and the caller tries to execute their request again.  Would there be a naming collision for the new column(s) if they still exist in the underlying table?

Reply:
Good catch! I'll update this step to handle if a column exists before creating it

Reply:
One option here might actually to handle this on the add column part.  Basically, if the column exists in BQ with the same type but doesn't exist in postgres yet, assume it's a rerun and don't fail?